### PR TITLE
Added support for optimised reference image PNGs

### PR DESCRIPTION
**Pull request by lukaskukacka**
**TL;DR: Don't fail test / comparison if recorded saved reference image PNG is optimized using tools like [ImageOptim](https://imageoptim.com/mac).**

## Overview

This allows to use apps like ImageOptim.app to optimise recorded reference images. Optimising reference image can heavily reduce its size, meaning we need to store less binary data in repo.

### Technical details

The core of this change is the way how `CGContextRef` are initialized.

#### Before (original code)
Each context was using `minBytesPerRow` to calculate buffer size. However, this was reason of the failure in case of optimized images. ImageOptim will for example reduce color space of the PNG from 32bit to 8bit (monochromatic) where possible. This results in `minBytesPerRow` being wrong value (not big enough to fit runtime rendered snapshot).

#### After (this change)
Contexts are always built based on "bigger" (the one needing more memory) context of the two images (saved reference vs rendered). This means context will always be built based on runtime rendered image. When rendering `CGContextDrawImage` to context, both runtime snapshot and loaded optimized reference images are "normalized" to the same context (colorspace etc.).

## Backwards compatibility

✅ 100% backwards compatible

Optimizing reference images is opt-in, optional and fully backwards compatible. This PR will not impact existing projects which don't use optimized reference images.

## Performance impact

✅ None (negligible)

- No **runtime** impact
    - Comparison code is not impacted.
    - Optimizations like `memcmp` works the same.
- No (negligible) **memory** impact
    - Rendering contexts to memory buffers is the same. Change Both contexts are mostly rendered using 4x8bits per pixel.
    - The only impact is the removal of the `minBytesPerRow` padding optimization, which should have almost no impact.

### CODE REVIEW

**Comment by lukaskukacka** [comment]
@reidmain is anyone still maintaining this project / any chance to get this merged please?

**Comment by lukaskukacka** [comment]
@reidmain Any chance for this to being merged? Is the project still being maintained please? PRs seems to be abandoned (this one is opened 10 months with no response)

**Review by reidmain — CHANGES_REQUESTED** [review]
I can no longer approve any changes to this repository. I'm going to request changes to this PR so it can be removed from my queue.

**Comment by lukaskukacka** [comment]
Thanks @alanzeino for having a look.

I believe both this and #140 have the same goal and almost identical implementation. This was just waiting and ready here 1.5y before #140 😀

Looking at the code the solution is almost the same as well.
The main difference seems to me that this PR uses color space based on reference image ( `CGImageGetColorSpace(contextConfigImage.CGImage)`) while #140 hardcodes color space (`CGColorSpaceCreateDeviceRGB()`), which I am not sure is safe to do.

**Comment by bernikovich** [comment]
Hey. I faced the same problem as the author and ended up with the same fix. The only concern is that it may fail when one image has more bits per component while the other has more bits per pixel and bytes per row. But this is probably very rare.
